File: src/data/Storage.ts
import type {
  Asset,
  Card,
  CardDraft,
  Deck,
  DeckSettings,
  DraftDecision,
  DraftResolution,
  Grade,
  ID,
  NewDraftInput,
  ProposalMetadata,
  ProposeDraftsResult,
  ReviewLog,
  SchedulerKind,
  SchedulingState,
  StartedStudy,
  StudyGradeOutcome,
  StudyRequest,
  StudyView,
} from '../domain/models'
import type { DeckBackup } from './backup'
import type { RepoSnapshot } from './sync/snapshot'
import type { DbOps } from './sync/merge'

/** Outcome of an import: deck names added fresh vs. names that replaced existing decks. */
export interface ImportResult {
  added: string[]
  replaced: string[]
}

export interface VersionedRepoSnapshot {
  snapshot: RepoSnapshot
  revision: number
}

export type ApplyMergeResult =
  | { status: 'applied'; revision: number }
  | { status: 'stale'; currentRevision: number }

/** Fields of a card that can be patched after creation. */
export interface CardPatch {
  front?: string
  back?: string
  tags?: string[]
  suspended?: boolean
  lastAgainAt?: number | null
  scheduling?: SchedulingState
}

export interface DeckPatch {
  name?: string
  color?: string
  settings?: DeckSettings
}

/** One atomic persisted review outcome. `fsrsGrade` is omitted for fixed steps. */
export interface ReviewCommit {
  cardId: ID
  deckId: ID
  patch: CardPatch
  reviewedAt: number
  fsrsGrade?: Grade
  daily?: { day: string; field: 'newIntroduced' | 'reviewsDone' }
}

/**
 * Native application port for collection persistence and study sessions.
 *
 * Packaged builds use SQLite and the shared Rust StudySession through
 * {@link ./TauriStorage}; browser tests inject the Dexie adapter without
 * changing feature code.
 */
export interface Storage {
  /** Re-run storage-backed UI queries after a mutation through this adapter. */
  subscribe(listener: () => void): () => void

  createDeck(name: string, kind?: SchedulerKind): Promise<Deck>
  listDecks(): Promise<Deck[]>
  getDeck(id: ID): Promise<Deck | undefined>
  deleteDeck(id: ID): Promise<void>
  updateDeck(id: ID, patch: DeckPatch): Promise<void>

  createCard(deckId: ID, front: string, back: string, tags?: string[]): Promise<Card>
  getCard(id: ID): Promise<Card | undefined>
  listCards(deckId: ID): Promise<Card[]>
  updateCard(id: ID, patch: CardPatch): Promise<void>
  deleteCard(id: ID): Promise<void>

  proposeDrafts(
    deckId: ID,
    inputs: NewDraftInput[],
    metadata: ProposalMetadata,
    dryRun?: boolean,
  ): Promise<ProposeDraftsResult>
  listDrafts(): Promise<CardDraft[]>
  resolveDraft(
    id: ID,
    expectedRevision: number,
    decision: DraftDecision,
  ): Promise<DraftResolution>

  startStudy(request: StudyRequest): Promise<StartedStudy>
  revealStudy(sessionId: string): Promise<StudyView>
  gradeStudy(sessionId: string, grade: Grade): Promise<StudyGradeOutcome>
  advanceStudyPreview(sessionId: string): Promise<StudyView>
  endStudy(sessionId: string): Promise<void>

  /** Atomically persist card/counter changes and an optional FSRS training event. */
  commitReview(commit: ReviewCommit): Promise<ReviewLog | null>
  listReviewLogs(deckId: ID): Promise<ReviewLog[]>

  /** Cards in a deck due at or before `now`, soonest-due first. */
  dueCards(deckId: ID, now: number): Promise<Card[]>
  /** How many cards in a deck are due at or before `now`. */
  countDue(deckId: ID, now: number): Promise<number>

  /** Today's cap counters for a deck; zeros when the day has no row yet. */
  getDailyStat(deckId: ID, day: string): Promise<{ newIntroduced: number; reviewsDone: number }>

  /** Insert decks+cards; any existing deck whose name matches an incoming deck
   *  is removed first (replace-by-name). IDs are regenerated. */
  importDecks(decks: DeckBackup[]): Promise<ImportResult>

  /** Full point-in-time snapshot and the local revision observed with it. */
  exportSnapshot(): Promise<VersionedRepoSnapshot>
  /** Apply a merge only if the local store is still at `expectedRevision`. */
  applyMerge(ops: DbOps, expectedRevision: number): Promise<ApplyMergeResult>

  // Assets (images/GIFs embedded in card markdown as asset:<hash>)
  putAsset(bytes: Uint8Array, mime: string): Promise<Asset>
  getAsset(hash: ID): Promise<Asset | undefined>
  sweepOrphanAssets(): Promise<void>
}

File: src-tauri/src/storage.rs
use rem_core::{
    ApplyMergeResult, Asset, Card, CardPatch, Collection, DbOps, Deck, DeckBackup, DeckPatch,
    ImportResult, ReviewCommit, ReviewLog, VersionedSnapshot,
};
use serde::Serialize;
use tauri::State;

fn message(error: rem_core::CollectionError) -> String {
    error.to_string()
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DailyCounts {
    new_introduced: u32,
    reviews_done: u32,
}

#[tauri::command]
pub fn storage_create_deck(
    collection: State<'_, Collection>,
    name: String,
    now: i64,
) -> Result<Deck, String> {
    collection.create_deck(&name, now).map_err(message)
}

#[tauri::command]
pub fn storage_list_decks(collection: State<'_, Collection>) -> Result<Vec<Deck>, String> {
    collection.list_decks().map_err(message)
}

#[tauri::command]
pub fn storage_get_deck(
    collection: State<'_, Collection>,
    id: String,
) -> Result<Option<Deck>, String> {
    collection.get_deck(&id).map_err(message)
}

#[tauri::command]
pub fn storage_delete_deck(
    collection: State<'_, Collection>,
    id: String,
    now: i64,
) -> Result<(), String> {
    collection.delete_deck(&id, now).map_err(message)
}

#[tauri::command]
pub fn storage_update_deck(
    collection: State<'_, Collection>,
    id: String,
    patch: DeckPatch,
    now: i64,
) -> Result<(), String> {
    collection.update_deck(&id, patch, now).map_err(message)
}

#[tauri::command]
pub fn storage_create_card(
    collection: State<'_, Collection>,
    deck_id: String,
    front: String,
    back: String,
    tags: Vec<String>,
    now: i64,
) -> Result<Card, String> {
    collection
        .create_card(&deck_id, &front, &back, tags, now)
        .map_err(message)
}

#[tauri::command]
pub fn storage_get_card(
    collection: State<'_, Collection>,
    id: String,
) -> Result<Option<Card>, String> {
    collection.get_card(&id).map_err(message)
}

#[tauri::command]
pub fn storage_list_cards(
    collection: State<'_, Collection>,
    deck_id: String,
) -> Result<Vec<Card>, String> {
    collection.list_cards(&deck_id).map_err(message)
}

#[tauri::command]
pub fn storage_update_card(
    collection: State<'_, Collection>,
    id: String,
    patch: CardPatch,
    now: i64,
) -> Result<(), String> {
    collection.update_card(&id, patch, now).map_err(message)
}

#[tauri::command]
pub fn storage_delete_card(
    collection: State<'_, Collection>,
    id: String,
    now: i64,
) -> Result<(), String> {
    collection.delete_card(&id, now).map_err(message)
}

#[tauri::command]
pub fn storage_commit_review(
    collection: State<'_, Collection>,
    commit: ReviewCommit,
) -> Result<Option<ReviewLog>, String> {
    collection.commit_review(commit).map_err(message)
}

#[tauri::command]
pub fn storage_list_review_logs(
    collection: State<'_, Collection>,
    deck_id: String,
) -> Result<Vec<ReviewLog>, String> {
    collection.list_review_logs(&deck_id).map_err(message)
}

#[tauri::command]
pub fn storage_due_cards(
    collection: State<'_, Collection>,
    deck_id: String,
    now: i64,
) -> Result<Vec<Card>, String> {
    collection.due_cards(&deck_id, now).map_err(message)
}

#[tauri::command]
pub fn storage_count_due(
    collection: State<'_, Collection>,
    deck_id: String,
    now: i64,
) -> Result<u64, String> {
    collection.count_due(&deck_id, now).map_err(message)
}

#[tauri::command]
pub fn storage_get_daily_stat(
    collection: State<'_, Collection>,
    deck_id: String,
    day: String,
) -> Result<DailyCounts, String> {
    collection
        .get_daily_stat(&deck_id, &day)
        .map(|(new_introduced, reviews_done)| DailyCounts {
            new_introduced,
            reviews_done,
        })
        .map_err(message)
}

#[tauri::command]
pub fn storage_import_decks(
    collection: State<'_, Collection>,
    decks: Vec<DeckBackup>,
    now: i64,
) -> Result<ImportResult, String> {
    collection.import_decks(decks, now).map_err(message)
}

#[tauri::command]
pub fn storage_export_snapshot(
    collection: State<'_, Collection>,
) -> Result<VersionedSnapshot, String> {
    collection.export_snapshot().map_err(message)
}

#[tauri::command]
pub fn storage_apply_merge(
    collection: State<'_, Collection>,
    operations: DbOps,
    expected_revision: u64,
    now: i64,
) -> Result<ApplyMergeResult, String> {
    collection
        .apply_merge(operations, expected_revision, now)
        .map_err(message)
}

#[tauri::command]
pub fn storage_put_asset(
    collection: State<'_, Collection>,
    bytes: Vec<u8>,
    mime: String,
    now: i64,
) -> Result<Asset, String> {
    collection.put_asset(&bytes, &mime, now).map_err(message)
}

#[tauri::command]
pub fn storage_get_asset(
    collection: State<'_, Collection>,
    hash: String,
) -> Result<Option<Asset>, String> {
    collection.get_asset(&hash).map_err(message)
}

#[tauri::command]
pub fn storage_sweep_orphan_assets(collection: State<'_, Collection>) -> Result<usize, String> {
    collection.sweep_orphan_assets().map_err(message)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn daily_counts_use_the_frontend_wire_shape() {
        let value = serde_json::to_value(DailyCounts {
            new_introduced: 2,
            reviews_done: 3,
        })
        .unwrap();
        assert_eq!(
            value,
            serde_json::json!({ "newIntroduced": 2, "reviewsDone": 3 })
        );
    }
}
